src/BotPlutusInterface/CoinSelection.hs

```diff
@@ -0,0 +1,479 @@
+--
+--     so, now the greedy search will select the following utxos:
+--     -- [10], [50], [210]
+--
+--     But, if we are using utxo vector with [210] then we don't need to
+--     consume vectors like [10] and [50].
+--     So, we filter such unnecessary vectors.
+--
+greedyApprox ::
+  forall (w :: Type) (effs :: [Type -> Type]).
+  Member (PABEffect w) effs =>
+  (Vector Integer -> Bool) ->
+  Vector Integer ->
+  Vector Integer ->
+  [Vector Integer] ->
+  Eff effs (Either Text [Int])
+greedyApprox stopSearch outVec txInsVec utxosVec
+  -- we stop the search if there are no utxos vectors left, as we will not be able to
+  -- select any further utxos as input to a transaction.
+  | null utxosVec =
+    printBpiLog @w (Debug [CoinSelectionLog]) "Greedy Pruning: The list of remanining UTxO vectors in null."
+      >> return (Right mempty)
+  -- we stop the search is the predicate `stopSearch` is true.
+  | stopSearch txInsVec =
+    printBpiLog @w (Debug [CoinSelectionLog]) "Greedy Pruning: Stopping search early."
+      >> return (Right mempty)
+  | otherwise =
+    runEitherT $ do
+      -- Here, we get the selected indexes of utxo vectors using greedy search.
+      selectedUtxosIdx <- newEitherT $ greedySearch @w stopSearch outVec txInsVec utxosVec
+
+      let -- Reverse the order of the selected vectors
+          -- The Idea here is that, the vectors that are selected at
+          -- last will have greater distance from the output vector.
+          -- Hence, they may contain all the values that's required for
+          -- the output vector.
+          revSelectedUtxosVec :: [Vector Integer]
+          revSelectedUtxosVec =
+            List.reverse $ selectedUtxosIdx ^.. folded . to (\idx -> utxosVec ^? ix idx) . folded
+
+          revSelectedUtxosIdx :: [Int]
+          revSelectedUtxosIdx = List.reverse selectedUtxosIdx
+
+      hoistEither $ loop txInsVec revSelectedUtxosIdx revSelectedUtxosVec
+  where
+    loop :: Vector Integer -> [Int] -> [Vector Integer] -> Either Text [Int]
+    loop newTxInsVec (idx : idxs) (vec : vecs) = do
+      -- Add the selected utxo vector to the current tx input vector.
+      newTxInsVec' <- addVec newTxInsVec vec
+
+      -- Get the old change vector
+      changeVec <- subVec outVec newTxInsVec
+
+      -- Get the new change vector
+      changeVec' <- subVec outVec newTxInsVec'
+
+      -- compare the distance between old change vector with output vector
+      -- and new change vector with the output vector.
+      case l2norm outVec changeVec' < l2norm outVec changeVec of
+        -- If the distance between new change vector and output
+        -- vector is smaller then we add that utxo vector
+        True -> (idx :) <$> loop newTxInsVec' idxs vecs
+        -- Else we check if we should stop the search here.
+        False | stopSearch newTxInsVec -> Right mempty
+        -- We add the current utxo vector.
+        False -> (idx :) <$> loop newTxInsVec' idxs vecs
```

Comment:
Could also use a wildcard here instead of `False`

Reply:
fixed